features/steps/thens/relations.py

```diff
@@ -189,108 +215,141 @@ def make_obj(s):
 
         accepted_values['applicable_entities'] = [s.split('/')[0] for s in make_obj(property_set_attr['applicable_entities'])]
 
-        accepted_values['applicable_type_values'] = property_set_attr.get('applicable_type_value', '').split(',')
+        # in the ifc4 data, predefined type restrictions are imposed as:
+        # | | | applicable_entities                                                     | | 
+        # | | | ['{entity}','{entity}/{predefinedtype}','{entity2}/{predefinedtype2}']  | |
+        accepted_values['predefined_types'] = [s.split('/')[1] for s in make_obj(property_set_attr['applicable_entities']) if '/' in s]
+
+        # in the ifc2x3 data, predefined type restrictions are imposed as:
+        # | | | applicable_type_value                     | | 
+        # | | | {entity}.PredefinedType={predefinedtype}  | |
+        if property_set_attr['applicable_type_value'] and '.PredefinedType=' in property_set_attr['applicable_type_value']:
+            accepted_values['predefined_types'].append(property_set_attr['applicable_type_value'].split('.PredefinedType=')[1])
+
+        # just in case convert to upper cae, probably not needed
+        accepted_values['predefined_types'] = sorted(set(s.upper() for s in accepted_values['predefined_types']))
+
+        # @todo this is actually wrong, we should consider pairs of <entity, predefined type> together
+        # and not separately. But in the majority of cases this will be good enough.
```

Comment:
I'm thinking of a way to handle this, e.g.
```
Scenario Outline:
Given an <IfcEntity>
Given its PropertySet
Then the value must be one of <X>
#or The value must be in <Y>

Examples:
Entity | X
EntityA | Value1,Value2,Value3
EntityB | Value4, Value5, Value6
EntityC | Value7, Value8, Value9

#or 
Entity A | AllowedPredefinedTypesX
Entity A | AllowedPredefinedTypesY
```
Perhaps at least for the ones for which errors are most likely to occur (to our knowledge) and exclude them from . 

Or we can do it the other way around and be more restrictive at 
 ```
   if 'IfcPropertySet Name attribute value must use predefined values according' in context.step.name:
        if name not in property_set_definitions.keys():
            yield ValidationOutcome(inst=inst, observed = {'value':name}, severity=OutcomeSeverity.ERROR)
```
and defined specific psets for values (in a separate csv) for the inverse of inst (which is a pset)? 
`context.model.get_inverse(inst)`

So for `IfcWall`, that would be one of the following values; 
https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcWall.htm#6.1.3.41.5-Property-sets

Reply:
I do like the idea to offload more of the python code to gherkin to make it more explicit. Especially the predefined type handling is now fully opaque in gherkin. It would be nice if we have a specific statement for that. Also filtering on whether the predefined type is required and/or present.

Reply:
Handled in https://github.com/buildingSMART/ifc-gherkin-rules/pull/192